lighthouse-plugin-publisher-ads/audits/cumulative-ad-shift.js

```diff
@@ -0,0 +1,170 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const i18n = require('lighthouse/lighthouse-core/lib/i18n/i18n');
+const {auditNotApplicable} = require('../messages/common-strings');
+const {Audit} = require('lighthouse');
+const {getScriptUrl} = require('../utils/network-timing');
+const {isAdIframe, isImplTag} = require('../utils/resource-classification');
+const {overlaps, toClientRect} = require('../utils/geometry');
+
+const UIStrings = {
+  title: 'Cumulative ad shift',
+  failureTitle: 'Reduce ad-related layout shift',
+  description:
+      'Measures [layout shifts](https://web.dev/cls) that were ' +
+          'caused by ads or happened near ads. Avoid layout shifts to improve ' +
+          'user experience.',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/**
+ * Audit to determine time for first ad request relative to page start.
+ */
+class CumulativeAdShift extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   * @override
+   */
+  static get meta() {
+    return {
+      id: 'cumulative-ad-shift',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      // @ts-ignore
+      scoreDisplayMode: Audit.SCORING_MODES.NUMERIC,
+      requiredArtifacts: ['traces', 'IFrameElements'],
+    };
+  }
+
+  /**
+   * @return {LH.Audit.ScoreOptions}
+   */
+  static get defaultOptions() {
+    // TODO tune this
+    return {
+      p10: 0.05,
+      median: 0.2,
+    };
+  }
+
+  /**
+   * @param {LH.TraceEvent} shiftEvent
+   * @param {Artifacts['IFrameElement'][]} ads
+   */
+  static isAdShift(shiftEvent, ads) {
+    if (!shiftEvent.args || !shiftEvent.args.data) {
+      return false;
+    }
+    for (const ad of ads) {
+      // Names come from external JSON
+      // eslint-disable-next-line camelcase
+      for (const node of shiftEvent.args.data.impacted_nodes || []) {
+        // eslint-disable-next-line camelcase
+        const /* number[] */ oldRect = node.old_rect || [];
+        const shiftRect = toClientRect(oldRect);
+        const adRect = ad.clientRect;
+        if (overlaps(shiftRect, adRect)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Computes the ad shift score for the page.
+   * @param {LH.TraceEvent[]} shiftEvents
+   * @param {Artifacts['IFrameElement'][]} ads
+   * @param {number} tagLoadTs
```

Comment:
Could you add a description of tagLoadTs here? Not obvious what "Ts" means. I'm guessing not TypeScript, but...?

Reply:
Timestamp. It's using the same abbreviation as chrome traces